File: project/backend-services/smart-irrigation-backend/domain-services/src/main/java/pt/sensae/services/smart/irrigation/backend/domainservices/data/DataMapper.java
package pt.sensae.services.smart.irrigation.backend.domainservices.data;

import pt.sensae.services.smart.irrigation.backend.domain.model.data.Data;
import pt.sensae.services.smart.irrigation.backend.domain.model.data.DataId;
import pt.sensae.services.smart.irrigation.backend.domain.model.data.Payload;
import pt.sensae.services.smart.irrigation.backend.domain.model.data.ReportTime;
import pt.sensae.services.smart.irrigation.backend.domain.model.data.payload.Illuminance;
import pt.sensae.services.smart.irrigation.backend.domain.model.data.payload.SoilMoisture;
import pt.sensae.services.smart.irrigation.backend.domain.model.device.DeviceId;
import pt.sharespot.iot.core.sensor.ProcessedSensorDataDTO;

public class DataMapper {

    public static Data dtoToModel(ProcessedSensorDataDTO dto) {
        var id = DataId.of(dto.dataId);
        var deviceId = DeviceId.of(dto.dataId);
        var reportedAt = ReportTime.of(dto.reportedAt);
        var lux = Illuminance.of(dto.data.illuminance.lux.floatValue());
        var moisture = SoilMoisture.of(dto.data.moisture.percentage.floatValue());
        var payload = new Payload(lux, moisture);
        return new Data(id, deviceId, reportedAt, payload);
    }
}

File: project/frontend-services/libs/smart-irrigation/mapper/src/lib/mappers/DataMapper.ts
import {SensorDataDTO} from "@frontend-services/smart-irrigation/dto";
import {Data} from "@frontend-services/smart-irrigation/model";
import {DeviceMapper} from "./DeviceMapper";
import {DataDetailsMapper} from "./DataDetailsMapper";

export class DataMapper {
  static dtoToModel(dto: SensorDataDTO): Data {
    return new Data(dto.dataId,
      DeviceMapper.dtoToModel(dto.device),
      new Date(Number(dto.reportedAt)),
      DataDetailsMapper.dataDtoToModel(dto.data, dto.device.type));
  }
}
